src/Microsoft.AspNet.Mvc.Core/DefaultControllerTypeProvider.cs

```diff
@@ -76,17 +80,46 @@ protected internal virtual bool IsController([NotNull] TypeInfo typeInfo)
             {
                 return false;
             }
-            if (typeInfo.Name.Equals("Controller", StringComparison.OrdinalIgnoreCase))
+            if (!typeInfo.Name.EndsWith(ControllerTypeName, StringComparison.OrdinalIgnoreCase) &&
+                !DerivesFromController(typeInfo, candidateAssemblies))
             {
                 return false;
             }
-            if (!typeInfo.Name.EndsWith("Controller", StringComparison.OrdinalIgnoreCase) &&
-                !typeof(Controller).GetTypeInfo().IsAssignableFrom(typeInfo))
+            if (typeInfo.IsDefined(typeof(NonControllerAttribute)))
             {
                 return false;
             }
 
             return true;
         }
+
+        private bool DerivesFromController(TypeInfo typeInfo, ISet<Assembly> candidateAssemblies)
+        {
+            // A type is a controller if it derives from a type that is either named "Controller" or has the suffix
+            // "Controller". We'll optimize the most common case of types deriving from the Mvc Controller type and
+            // walk up the object graph if that's not the case.
+            if (ControllerTypeInfo.IsAssignableFrom(typeInfo))
+            {
+                return true;
+            }
+
+            while (typeInfo != ObjectTypeInfo)
+            {
+                var baseTypeInfo = typeInfo.BaseType.GetTypeInfo();
+                // A base type will be treated as a controller if
+                // a) it ends in the term "Controller" and
+                // b) it's assembly is one of the candidate assemblies we're considering. This ensures that the assembly
+                // the base type is declared in references Mvc.
+                if (baseTypeInfo.Name.EndsWith(ControllerTypeName, StringComparison.Ordinal) &&
```

Comment:
Not sure I get this concern, this seems like something that will never happen, as it's a feature on a feature on a feature.